### ISSUE TITLE
Signing Commits with GPG

### ISSUE THREAD

**Issue by jwillikers** [flathub/com.sublimemerge.App#18]
I have configured Git to sign commits by default.

```sh
$ git config --global commit.gpgSign true
```

When I use the Sublime Merge Flatpak, I can't commit any changes because signing fails.
This creates the following error:

```sh
error: gpg failed to sign the data
fatal: failed to write commit object
```

This is not an issue using the version of Sublime Merge installed from it's official repositories for Linux.

**Comment by ptomato** [comment]
Do you know what would be needed to be able to sign the commit? The `gpg` binary is available from within the sandbox, so it's failing for some other reason.

**Comment by fariasmarcos** [comment]
I think i just found out what is missing. It seems to be pinentry. Without it, gnupg cannot prompt the user for a keyring password, then proceeds to fail.

```
[📦 com.sublimemerge.App ~]$ echo "test" | gpg --clearsign
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

test
gpg: clear-sign failed: No pinentry
gpg: [stdin]: clear-sign failed: No pinentry
```

**Comment by ptomato** [comment]
Thanks! Would you be able to try the test build in https://github.com/flathub/com.sublimemerge.App/pull/19 and see if it solves your problem?

**Comment by fariasmarcos** [comment]
The error still happens :(
```
[📦 com.sublimemerge.App ~]$ echo "test" | gpg --clearsign
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

test
gpg: clear-sign failed: No pinentry
gpg: [stdin]: clear-sign failed: No pinentry
```

However, pinentry is now available from within the sandbox, and shows the standard message when executed directly
```
[📦 com.sublimemerge.App ~]$ pinentry
OK Pleased to meet you
```

Everything else seems to be ok, just the connection between gpg-agent and pinentry is broken.

```
❯ flatpak info com.sublimemerge.App

Sublime Merge - Sublime Merge is a Git client, from the makers of Sublime Text

        ID: com.sublimemerge.App
       Ref: app/com.sublimemerge.App/x86_64/test
      Arq.: x86_64
      Ramo: test
    Versão: 0.2039
   Licença: LicenseRef-proprietary=https://www.sublimetext.com/eula
    Origem: app-origin
   Coleção: org.flathub.Stable.Build35096
Instalação: system
 Instalado: 15,8 MB
   Runtime: org.freedesktop.Sdk/x86_64/20.08
       Sdk: org.freedesktop.Sdk/x86_64/20.08

    Commit: b94a84169c87ed4c18d24feff26a986bf2bb7e553ce4fa5683f2a87e7e87305a
   Assunto: Add pinentry (37df0255)
      Data: 2021-01-11 05:32:41 +0000
```

**Comment by jwillikers** [comment]
@ptomato Thanks for working on this!
I think the following command can fix this by setting the appropriate permissions.

```sh
$ flatpak override --user --filesystem=xdg-run/gnupg:ro com.sublimemerge.App
```

This is how the Thunderbird Flatpak accesses GnuPG as seen [here](https://github.com/flathub/org.mozilla.Thunderbird/blob/e68475dd84086ac1c0b10f9e2c182b5a14c11710/org.mozilla.Thunderbird.json#L39).
The Thunderbird Flatpak used to bundle pinentry, but no longer does now so I don't think it needs to be bundled for the Sublime Merge Flatpak either.

**Comment by ptomato** [comment]
Very helpful, thanks.
There should be another test build in #19 that you can try now.

**Comment by jwillikers** [comment]
I can't get that build to work for some reason on my Fedora / GNOME systems. I think I was able to get this working on my Pinebook Pro running KDE, so I'll test again on that later.

**Comment by fariasmarcos** [comment]
The new test build now works, but only if i trigger a sign command on my host machine, and type the private key's password, before try to use it inside the sandbox. If i don't, it complains about failed IPC calls, see:
```
[📦 com.sublimemerge.App ~]$ echo test | gpg --clearsign
gpg: can't connect to the agent: Chamada de conexão de IPC falhada
gpg: can't connect to the agent: Chamada de conexão de IPC falhada
gpg: keydb_search failed: Nenhum agente em execução
gpg: no default secret key: Nenhum agente em execução
gpg: [stdin]: clear-sign failed: Nenhum agente em execução
```
git from within sublime merge will throw this error message:
```
error: gpg failed to sign the data
fatal: failed to write commit object
```
After sign anything on the host with the same private key, it works flawlessly!

My host machine runs Fedora Workstation 33 and Flatpak 1.8.4.

**Comment by fariasmarcos** [comment]
Good news @ptomato! I just fired up a Debian 10 VM, installed the build test, and had no problems whatsoever.
As soon as i made a test commit, it launched pinentry on the host OS (in this case, Debian) and a signed commit was created!

Fedora seems to be the one to blame for the odd behaviour, maybe something related to SELinux or something i messed up? I can't tell.

**Comment by jwillikers** [comment]
Yep, works just fine on Manjaro KDE. There is still a Fedora-specific issue, but it is still related to the Flatpak because I am able to sign my commits using Sublime Merge installed from the official RPM repository.

**Comment by ptomato** [comment]
Hmm. I guess we should not merge this yet, then, but I have no idea what could be causing the difference.

**Comment by jwillikers** [comment]
I've tested the pinentry build and added the necessary permission manually, but the same error happens there on Fedora 33.

I'll look more into it today and see what I can find out by signing commits from within the Flatpak's sandbox.

**Comment by jwillikers** [comment]
Okay, testing today on a different Fedora 33 workstation I was able to sign commits no problem with the Flatpak from the PR.
Just to be certain, I confirmed the Flatpak from the stable branch still fails to sign commits and then tested that the updated build does indeed work.
Perhaps my manual tinkering with permissions of the Flatpak caused the issue.
Everything looks good now.
Sorry about the hold up and thanks for all the help!

**Comment by ptomato** [comment]
OK, I wish I knew what was going on with your other system, but I'll merge the pull request now. Thanks both of you for the help.

**Comment by jwillikers** [comment]
So, I've had this issue reappear in Fedora Silverblue 34. I checked the SELinux logs, and that isn't logging anything about this. I have a hunch that this can be fixed by adding one or both of  `--talk-name=org.freedesktop.secrets` and `--filesystem=xdg-run/keyring`, which gitg uses as seen [here](https://github.com/flathub/org.gnome.gitg/blob/03a4f570e190edab7b675f64039adfe21a1af600/org.gnome.gitg.json#L21). I'm currently testing the former having used the command `sudo flatpak override --talk-name=org.freedesktop.secrets com.sublimemerge.App` to enable it.
It appears to be working so far, but since this might be flaky, I'll see if the problem no longer occurs with this option over the next couple of days.

**Comment by jwillikers** [comment]
Using deduction and the fact that the VSCodium Flatpak can sign commits, it is almost certainly `--talk-name=org.freedesktop.secrets` which is required.
The VSCodium Flatpak uses ` --talk-name=org.freedesktop.secrets`, as seen [here](https://github.com/flathub/com.vscodium.codium/blob/54363f700aae22284ac9b972723d60cd6625c823/com.vscodium.codium.yaml#L18), but not `--filesystem=xdg-run/keyring`.

**Comment by jwillikers** [comment]
I've figured this one out. Long story short, both `--talk-name=org.freedesktop.secrets` and `--filesystem=xdg-run/keyring` are necessary. The confusion was due to the fact that signing commits fails in Sublime Merge until the GPG key is unlocked in GNOME keyring, such as after signing a commit on the command-line. VSCodium actually has the exact same issue as Sublime Merge does here, but this just worked in the Gitg Flatpak. I've submitted PR #24 to fix this.
